ReadMe.txt
/*
 * This readme explains how to run the voting protocol on Ethereum.
 * by Digendra Rai.
 *
*/


Running the source code requires the following installations on Ubuntu.
	- geth 1.7.3-stable
	- nodejs v8.9.1
	- setting the geth environment variables in ubuntu

Step 1: After the installations gets completed, in the terminal the following command line should be typed for starting the geth instance at specific port.
	--> geth --datadir . --networkid 71 --port 30304 --rpcport 8650 --rpc --rpccorsdomain "*" console

		Note: For each of the nodes the --rpcport and --port should be different. If many nodes are to be connected and run.
	
Step 2: The nodejs instance should be run next using the terminal. This can be done by simply typring nodejs in the console.

Step 3: The following steps are to be followed in the nodejs command promt for deploying the smart contract files in the geth instance. All the smart contract files are solidity supported files.

		step a: Web3 = require('web3'). // If web3 id not installed install it.

		step b: web3 = new Web3(new Web3.providers.HttpProvider("http://localhost:8545")). // The port number should be different for different geth instance nodes.

		step c: code = fs.readFileSync('Polling.sol').toString()

		step d: solc = require('solc'). //Install solidity compiler which is called "solc", if not installed.

		step e:  compiledCode = solc.compile(code)

		abiDefinition = JSON.parse(compiledCode.contracts[':Polling'].interface)

		step f: VotingContract = web3.eth.contract(abiDefinition)

		stpe g: byteCode = compiledCode.contracts[':Voting'].bytecode

		step h: deployedContract = VotingContract.new(['Rama','Nick','Jose'],{data: '0x'.concat( byteCode), from: web3.eth.accounts[0], gas: 4700000}). // This will deploy the smart contract on the geth instance blockchain.

The following steps explains the steps to talk to the smart contract function which is on the geth instance blockchain. It's optional and is used to check whether parameters passed to the smart function works or not by calling the function.

		step i: contractInstance = VotingContract.at(deployedContract.address)

		step j: contractInstance.totalVotesFor.call('Rama').

		step k: contractInstance.voteForCandidate('Rama', {from: web3.eth.accounts[0]})


After the above three steps are followed, for connecting mutiple geth nodes requires the following steps:

	At each of the geth instance command promt check the instance address and port by typing the following command on the promt;
			
			admin.nodeInfo.encode().

	Then, copy paste the nodes instance address and port in the other nodes geth intance command promt using the following command.

		admin.addPeer("node Instance and port");

	use the: "miner.start(); admin.sleepBlocks(7); admin.stop()" in any one of the node's geth instance commmand promt. This will take some time for syncing the multiple nodes.

	For addition of the geth nodes requires the above two commands to be followed.


Now, open "Register.html" from the PollingTmp folder in your browser and you are ready to vote.

NOTE: After the publishing of commitments by clicking the button, it's necessary that command "miner.start(); admin.sleepBlocks(7); admin.stop()" should be run on any node's geth instance.

NOTE: After the voting is done againg the command "miner.start(); admin.sleepBlocks(7); admin.stop()" should be run on any of the node's geth instance.



PollingTmp/history
web3.eth.accounts
eth.getBalance(web3.eth.accounts[0])
miner.start();admin.sleepBlocks(7);miner.stop();
eth.getBalance(web3.eth.accounts[0])
miner.start();admin.sleepBlocks(7);miner.stop();
eth.getBalance(web3.eth.accounts[0])
miner.start();admin.sleepBlocks(7);miner.stop();

PollingTmp/testZKF.js
Web3 = require('web3');
web3 = new Web3(new Web3.providers.HttpProvider("http://localhost:8545"));

var ZKFContract = web3.eth.contract([{"constant":true,"inputs":[],"name":"Tree","outputs":[{"name":"","type":"address"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":false,"inputs":[{"name":"_input","type":"uint256[]"},{"name":"_user","type":"address"},{"name":"merkelTreeContractAddress","type":"address"}],"name":"checkMerkelTree","outputs":[{"name":"ck","type":"bool"}],"payable":false,"stateMutability":"nonpayable","type":"function"},{"constant":false,"inputs":[{"name":"_data","type":"uint256"},{"name":"_user","type":"address"},{"name":"merkelTreeContractAddress","type":"address"}],"name":"addIntoMerkelTree","outputs":[{"name":"success","type":"bool"}],"payable":false,"stateMutability":"nonpayable","type":"function"},{"constant":true,"inputs":[{"name":"p","type":"uint256"},{"name":"q","type":"uint256"},{"name":"g","type":"uint256"},{"name":"zkf","type":"uint256[4]"}],"name":"validateZKF","outputs":[{"name":"res","type":"uint256"}],"payable":false,"stateMutability":"view","type":"function"}])

var ballot= ZKFContract.new(
	{
		from: web3.eth.accounts[0],
		data: '0x6060604052341561000f57600080fd5b6103a88061001e6000396000f3006060604052600436106100485763ffffffff60e060020a60003504166368d82fb7811461004d578063730711da1461007c5780639ce5812b146100f1578063a76b29d814610119575b600080fd5b341561005857600080fd5b61006061016f565b604051600160a060020a03909116815260200160405180910390f35b341561008757600080fd5b6100dd600460248135818101908301358060208181020160405190810160405280939291908181526020018383602002808284375094965050600160a060020a0385358116956020013516935061018392505050565b604051901515815260200160405180910390f35b34156100fc57600080fd5b6100dd600435600160a060020a0360243581169060443516610271565b341561012457600080fd5b61015d6004803590602435906044359060e4606482608060405190810160405291908282608080828437509395506102f9945050505050565b60405190815260200160405180910390f35b6000546101009004600160a060020a031681565b6000805474ffffffffffffffffffffffffffffffffffffffff001916610100600160a060020a03848116820292909217808455041663276ac4e28585846040516020015260405160e060020a63ffffffff8516028152600160a060020a038216602482015260406004820190815290819060440184818151815260200191508051906020019060200280838360005b8381101561022a578082015183820152602001610212565b505050509050019350505050602060405180830381600087803b151561024f57600080fd5b6102c65a03f1151561026057600080fd5b505050604051805195945050505050565b6000805474ffffffffffffffffffffffffffffffffffffffff001916610100600160a060020a03848116820292909217808455041663ee35f1638585846040516020015260405160e060020a63ffffffff85160281526004810192909252600160a060020a03166024820152604401602060405180830381600087803b151561024f57600080fd5b6000808080808551935060208601519250604086015191506060860151905061034f61034961033161032b8a8761035c565b8c610361565b61034461033e868661035c565b8d610361565b610378565b8a610361565b9998505050505050505050565b900a90565b600080828481151561036f57fe5b06949350505050565b02905600a165627a7a7230582091a900b1a01e1aa324314350ca934cd83d4aed5b60a524a0b6c74033bfa27ef10029',
		gas: '2700000'
	},
	function(e,contract){
		console.log(e,contract);
		if (typeof contract.address !== 'undefined') {
			console.log('contract mined at: ' +contract.address);
		}
	}
)
//var deployedZKFAddress= '0x73694ebfd9ee9816be02873268708abf1e94d48a';
//var ZKFInstance = ZKFContract.at(deployedZKFAddress);

//var merkleTreeContractAddress='0x5f155fbe1a0124eff716592af46d7a09ffd25fad';




PollingTmp/nodeTest.js

Web3 = require('web3');
web3 = new Web3(new Web3.providers.HttpProvider("http://localhost:8545"));

var ZKFContract = web3.eth.contract([{"constant":true,"inputs":[],"name":"Tree","outputs":[{"name":"","type":"address"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":false,"inputs":[{"name":"_input","type":"uint256[]"},{"name":"_user","type":"address"},{"name":"merkelTreeContractAddress","type":"address"}],"name":"checkMerkelTree","outputs":[{"name":"ck","type":"bool"}],"payable":false,"stateMutability":"nonpayable","type":"function"},{"constant":false,"inputs":[{"name":"_data","type":"uint256"},{"name":"_user","type":"address"},{"name":"merkelTreeContractAddress","type":"address"}],"name":"addIntoMerkelTree","outputs":[{"name":"success","type":"bool"}],"payable":false,"stateMutability":"nonpayable","type":"function"},{"constant":true,"inputs":[{"name":"p","type":"uint256"},{"name":"q","type":"uint256"},{"name":"g","type":"uint256"},{"name":"zkf","type":"uint256[]"}],"name":"validateZKF","outputs":[{"name":"res","type":"uint256"}],"payable":false,"stateMutability":"view","type":"function"}])

var deployedZKFAddress= '0xb0b9b5d037f98c16d42b3de96741e5dde0d39af4';
var ZKFInstance = ZKFContract.at(deployedZKFAddress);

var merkleTreeContractAddress='0xac187394927c071645f800ce549a9446c0756b80';

var Chance= require('chance')
var hash = require('hash.js')

chance= new Chance(Math.random);

var secretSN = chance.integer({min:0, max: 100})
var secretKey = chance.integer({min:0, max: 100})
console.log("secretSN:",secretSN)
console.log("Secret Key:",secretKey)

var commitment= '0x'.concat(hash.sha256().update(secretKey).digest('hex'))

console.log("Hash Secret SN:",commitment)

//var putCommitmentOnBC= ZKFInstance.addIntoMerkelTree(commitment,web3.eth.accounts[2],merkleTreeContractAddress,{from:web3.eth.accounts[2]})

//console.log("putCommitmentOnBC:", putCommitmentOnBC)

//var success= ZKFInstance.checkMerkelTree.call([1,commitment],web3.eth.accounts[2],merkleTreeContractAddress)


//console.log(success);


var acct= web3.eth.accounts[2]
var acctBal= web3.fromWei(web3.eth.getBalance(acct),"ether")
var balance= parseInt(acctBal)

console.log("Before Balance:",acct)

//web3.eth.sendTransaction({from:web3.eth.accounts[3], to:web3.eth.accounts[2], value: web3.toWei(1, "ether")})

//var acctBal= web3.fromWei(web3.eth.getBalance(acct),"ether")
//var balance= parseInt(acctBal)

//console.log("After Balance:",balance)


/***************************
	This is for Polling
*******************************/
	var test= true

	var poll= function() {
		return test
	}


/******
	This is for ZKF
	*******************************************************/

var bigInt = require("big-integer");

var primeQ= bigInt(14)
var primeP= bigInt(primeQ).multiply(bigInt(2).add(1))

console.log("PrimeP:",primeP.toString())

var g= bigInt(8)
var cm= bigInt(8)
var x= bigInt(10)


var another= require('./ZKF.js')
var makeZKF= another(primeP,primeQ,g,cm,x) 

console.log("MakeZKF:",makeZKF)

var arr= [4]
arr[0]= makeZKF[0]
arr[1]= makeZKF[1]
arr[2]= makeZKF[2]
arr[3]= makeZKF[3]

var q= bigInt(arr[1]).abs().toString()

var checkZ= require('./checkZKF.js')
console.log("RES",checkZ(primeP,primeQ,g,arr))

var validZKF= ZKFInstance.validateZKF.call(primeP.toString(),primeQ.toString(),g.toString(),[arr[0],q,arr[2],arr[3]])

console.log("ValidZKF:",validZKF.toString())

module.exports = {
	poll: poll
}




PollingTmp/AuditVote.js
Web3= require('web3');
web3 = new Web3(new Web3.providers.HttpProvider("http://localhost:8545"));

var AuditVoteContract = web3.eth.contract([{"constant":true,"inputs":[{"name":"_a","type":"bytes32"},{"name":"_b","type":"bytes32"}],"name":"hashTwo","outputs":[{"name":"hashed","type":"bytes32"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[{"name":"_data","type":"uint256"}],"name":"checkVotesData","outputs":[{"name":"complete","type":"bool"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":false,"inputs":[{"name":"_input","type":"uint256"}],"name":"addVotesData","outputs":[{"name":"success","type":"uint256"}],"payable":false,"stateMutability":"nonpayable","type":"function"}])

/*var ballot= AuditVoteContract.new(
	{
			from: web3.eth.accounts[0],
			data: '0x6060604052341561000f57600080fd5b6102008061001e6000396000f3006060604052600436106100565763ffffffff7c01000000000000000000000000000000000000000000000000000000006000350416631c8e8d7a811461005b578063221ec24214610086578063ceae4930146100b0575b600080fd5b341561006657600080fd5b6100746004356024356100c6565b60405190815260200160405180910390f35b341561009157600080fd5b61009c6004356100e9565b604051901515815260200160405180910390f35b34156100bb57600080fd5b610074600435610168565b600082826040519182526020820152604090810190518091039020905092915050565b6000806000808460405190815260200160405190819003902060008681526020818152604080832084845260010190915290205490925060ff161561013c5761013282846100c6565b9050809250610145565b60009350610160565b60008581526020819052604090205483141561013c57600193505b505050919050565b60008060008084604051908152602001604051908190039020600086815260208190526040902054909350915061019f83836100c6565b60008681526020818152604080832096835260018088018352908320805460ff191682179055978252529092555091929150505600a165627a7a723058204703c05011c7ffa557d554c8e69f9dec05704d8a672fedac6ce4d46d3b81d6dc0029',
			gas: '2700000'
		},
		function(e,contract){
			console.log(e,contract);
			if (typeof contract.address !== 'undefined') {
				console.log('contract mined at: ' +contract.address);
			}
		}
	)

*/

var AuditVoteAddress= "0x99f8f4ddcc851dea23936355efcfd368b514f3c7"
var AuditVoteInstance= AuditVoteContract.at(AuditVoteAddress)


//var makeTree= AuditVoteInstance.addVotesData(17,{from:web3.eth.accounts[0]})
//console.log("makeTree: ",makeTree)

var checkTree= AuditVoteInstance.checkVotesData.call(18).toString()
console.log("checkTree: ",checkTree)








PollingTmp/MerkleTree.js
Web3 = require('web3');
web3 = new Web3(new Web3.providers.HttpProvider("http://localhost:8545"));

var MerkleTreeContract = web3.eth.contract([{"constant":true,"inputs":[{"name":"_a","type":"bytes32"},{"name":"_b","type":"bytes32"}],"name":"hashTwo","outputs":[{"name":"hashed","type":"bytes32"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[{"name":"_data","type":"uint256[]"},{"name":"_user","type":"address"}],"name":"checkDataIntegrity","outputs":[{"name":"complete","type":"bool"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[{"name":"_user","type":"address"}],"name":"getUserRoot","outputs":[{"name":"root","type":"bytes32"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":false,"inputs":[],"name":"newUser","outputs":[],"payable":false,"stateMutability":"nonpayable","type":"function"},{"constant":true,"inputs":[{"name":"","type":"address"}],"name":"users","outputs":[{"name":"root","type":"bytes32"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":false,"inputs":[{"name":"_input","type":"uint256"},{"name":"_user","type":"address"}],"name":"addData","outputs":[{"name":"success","type":"uint256"}],"payable":false,"stateMutability":"nonpayable","type":"function"},{"constant":true,"inputs":[],"name":"empty","outputs":[{"name":"","type":"bytes32"}],"payable":false,"stateMutability":"view","type":"function"}])

var ballot= MerkleTreeContract.new(
	{
		from: web3.eth.accounts[0],
		data: '0x6060604052341561000f57600080fd5b6103648061001e6000396000f3006060604052600436106100825763ffffffff7c01000000000000000000000000000000000000000000000000000000006000350416631c8e8d7a8114610087578063276ac4e2146100b257806363c8baad146101205780636cebad981461013f578063a87430ba14610154578063ee35f16314610173578063f2a75fe414610195575b600080fd5b341561009257600080fd5b6100a06004356024356101a8565b60405190815260200160405180910390f35b34156100bd57600080fd5b61010c600460248135818101908301358060208181020160405190810160405280939291908181526020018383602002808284375094965050509235600160a060020a031692506101cb915050565b604051901515815260200160405180910390f35b341561012b57600080fd5b6100a0600160a060020a036004351661027c565b341561014a57600080fd5b610152610297565b005b341561015f57600080fd5b6100a0600160a060020a03600435166102ab565b341561017e57600080fd5b6100a0600435600160a060020a03602435166102bd565b34156101a057600080fd5b6100a0610332565b600082826040519182526020820152604090810190518091039020905092915050565b600080548180805b865183101561025d578683815181106101e857fe5b90602001906020020151604051908152602001604051908190039020600160a060020a0387166000908152600160208181526040808420858552909201905290205490925060ff16156102495761023f82856101a8565b9050809350610252565b60009450610272565b6001909201916101d3565b6102668661027c565b81141561024957600194505b5050505092915050565b600160a060020a031660009081526001602052604090205490565b600160a060020a0333166000526001602052565b60016020526000908152604090205481565b60008060008085604051908152602001604051809103902092506102e08561027c565b91506102ec83836101a8565b600160a060020a0395909516600081815260016020818152604080842097845287830182528320805460ff19168317905592909152908190529490925550919392505050565b600054815600a165627a7a723058201ee55b1eb82e966fae213f385bfa3afb5974654b9031d31ade3368c63b88395f0029',
		gas: '2700000'
	},
	function(e,contract){
		console.log(e,contract);
		if (typeof contract.address !== 'undefined') {
			console.log('contract mined at: ' +contract.address);
		}
	}
)
//var deployedZKFAddress= '0x73694ebfd9ee9816be02873268708abf1e94d48a';
//var ZKFInstance = ZKFContract.at(deployedZKFAddress);

//var merkleTreeContractAddress='0x5f155fbe1a0124eff716592af46d7a09ffd25fad';




PollingTmp/testCrypto.js
Web3= require('web3');

web3 = new Web3(new Web3.providers.HttpProvider("http://localhost:8650"));

var candidateNames= ["Rama","Nick","Jose"];
var candidates = {"Rama": "candidate-1", "Nick": "candidate-2", "Jose": "candidate-3"}

var PollingContract = web3.eth.contract([{"constant":true,"inputs":[{"name":"candidate","type":"bytes32"}],"name":"totalVotesFor","outputs":[{"name":"","type":"bytes32[]"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[{"name":"","type":"bytes32"}],"name":"votesReceived","outputs":[{"name":"","type":"uint256"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[{"name":"","type":"bytes32"},{"name":"","type":"uint256"}],"name":"votersEncryptedVotes","outputs":[{"name":"","type":"bytes32"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":false,"inputs":[{"name":"candidate","type":"bytes32"},{"name":"encryptedVotes","type":"bytes32"}],"name":"voteForCandidate","outputs":[],"payable":true,"stateMutability":"payable","type":"function"},{"constant":true,"inputs":[{"name":"","type":"uint256"}],"name":"candidateList","outputs":[{"name":"","type":"bytes32"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[{"name":"","type":"address"}],"name":"voterInfo","outputs":[{"name":"voterAddress","type":"address"},{"name":"voted","type":"bool"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[{"name":"candidate","type":"bytes32"}],"name":"indexOfCandidate","outputs":[{"name":"","type":"uint256"}],"payable":false,"stateMutability":"view","type":"function"},{"inputs":[{"name":"candidateNames","type":"bytes32[]"}],"payable":false,"stateMutability":"nonpayable","type":"constructor"}])


var pollingContractAddress= '0x1d3b190e4fbc8bca54d573414da366dcd28846a0'
var pollingInstance= PollingContract.at(pollingContractAddress)

var aesjs = require("aes-js");

function voteForCandidate() {

  	candidateName = $("#candidate").val();
	//alert("candidateName: "+candidateName)

	if (candidateName.length != 0){

		var vote1or0= ChooseVote();
		//alert("VOTES: "+vote1or0)

		if(vote1or0 != 7){
			
			var encryptedVote= aesEncryption(vote1or0)
			alert("EncryptedVote: "+encryptedVote)

			var ok= pollingInstance.voteForCandidate(candidateName, encryptedVote, {from: web3.eth.accounts[0]});
			alert("OK: "+ok)

			CountDown(5)
		}
		else
			alert("Yes or No Vote not clicked")
	
	}
	else
		alert("Candidate Name is NULL")
}

function ChooseVote() {
    //document.getElementById("").value = "Johnny Bravo";

	var voteRes;
	if(document.getElementById('zeroVote').checked){
		//alert(document.getElementById('zeroVote').value)
		voteRes= document.getElementById('zeroVote').value
	}
	else if(document.getElementById('oneVote').checked){
		//alert(document.getElementById('oneVote').value)
		voteRes= document.getElementById('oneVote').value
	}
	else{
		voteRes= 7
	}
	//alert("Vote0or1: "+voteRes)
	return voteRes
}

function totalVotesFor(){


	candidateNames = Object.keys(candidates);
  	for (var i = 0; i < candidateNames.length; i++) {
    	let name = candidateNames[i];

		var val= countVotesFor(name)
    	$("#" + candidates[name]).html(val);
  	}

}


function countVotesFor(name){

	var orgString= []
	var decryptedVote= []

	var val= pollingInstance.totalVotesFor.call(name).toString()
	//alert(val)
	
	var str_array = val.split(',');

	for(var i = 0; i < str_array.length; i++) {
		// Trim the excess whitespace.
		str_array[i] = str_array[i].replace(/^\s*/, "").replace(/\s*$/, "");
		//alert("Splitted String: "+str_array[i]);

		orgString.push(web3.toAscii(str_array[i]))	
	}
	
	for(var i = 0; i < str_array.length; i++){

		//alert("OrgString: "+orgString[i])
		decryptedVote[i]= aesDecryption(orgString[i]);
	}

	var countVotes= 0;
	for(var j= 0; j < decryptedVote.length; j++){
		
		if(decryptedVote[j] == 'one')
			countVotes += 1;
	}
	//alert("VotesCounted: "+countVotes)

	return countVotes
}

function cleanArray(actual) {
  var newArray = new Array();
  for (var i = 0; i < actual.length; i++) {
    if (actual[i]) {
      newArray.push(actual[i]);
    }
  }
  return newArray;
}


function showToButtonForVotesSummary(){

	var votesCountButton= document.getElementById("votesSumm").disabled= false;
}

function aesEncryption(text){

		// An example 128-bit key
	var key = [ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16 ];

	// The initialization vector (must be 16 bytes)
	var iv = [ 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,35, 36 ];

	// Convert text to bytes
	var textBytes = aesjs.utils.utf8.toBytes(text);

	var aesOfb = new aesjs.ModeOfOperation.ofb(key, iv);
	var encryptedBytes = aesOfb.encrypt(textBytes);

	//alert("EncryptedBytes: "+encryptedBytes)

	// To print or store the binary data, you may convert it to hex
	var encryptedHex = aesjs.utils.hex.fromBytes(encryptedBytes);
	//alert("EncryptedHex: "+encryptedHex);

	return encryptedHex
}

function aesDecryption(encryptedHex){

	var key = [ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16 ];
	// The initialization vector (must be 16 bytes)
	var iv = [ 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,35, 36 ];

	// When ready to decrypt the hex string, convert it back to bytes
	var encryptedBytes = aesjs.utils.hex.toBytes(encryptedHex);

	var cleanedEncryptedBytes= cleanArray(encryptedBytes);
	//alert("Cleaned Array: "+cleanedEncryptedBytes)

	// The output feedback mode of operation maintains internal state,
	// so to decrypt a new instance must be instantiated.
	var aesOfb = new aesjs.ModeOfOperation.ofb(key, iv);
	var decryptedBytes = aesOfb.decrypt(cleanedEncryptedBytes);
	//alert('hi: '+decryptedBytes)

	// Convert our bytes back into text
	var decryptedText = aesjs.utils.utf8.fromBytes(decryptedBytes);

	//alert("Decrypt: "+decryptedText);
	return decryptedText;
}



function CountDown(time){

	var interval= setInterval(function() {
		document.getElementById("demo").innerHTML= --time
		//document.getElementById("voteID").disabled= true

		if(time <= 0){

			document.getElementById("demo").innerHTML= "YOU Reaaady"
			//document.getElementById("voteID").disabled= false
			clearInterval(interval)
			
			x.style.display = "none";
		}
	
	},1000)
	
	//document.getElementById("voteID").disabled= true

	//setInterval(function(){Timer()},1000)
	//var myVar= setTimeout(alertButton,3000)
	
}


PollingTmp/polling.js
/*
 * This js File provides the Functionality of:
 * 		- Voting for a candidate
 * 		- Checking if the Vote is put on the Blockchain which is used later for counting.
 *		- AES Encryption and Decryption of Votes.
 *		- Vote Summary for all the Candidates.
*/


Web3= require('web3');
web3 = new Web3(new Web3.providers.HttpProvider("http://localhost:8650"));

var candidateNames= ["Rama","Nick","Jose"];
var candidates = {"Rama": "candidate-1", "Nick": "candidate-2", "Jose": "candidate-3"}

var PollingContract = web3.eth.contract([{"constant":true,"inputs":[{"name":"encryptedVotes_data","type":"bytes32"}],"name":"checkVotesData","outputs":[{"name":"complete","type":"bool"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[{"name":"candidate","type":"bytes32"}],"name":"totalVotesFor","outputs":[{"name":"","type":"bytes32[]"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[{"name":"","type":"bytes32"}],"name":"votesReceived","outputs":[{"name":"","type":"uint256"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[{"name":"","type":"bytes32"},{"name":"","type":"uint256"}],"name":"votersEncryptedVotes","outputs":[{"name":"","type":"bytes32"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":false,"inputs":[{"name":"candidate","type":"bytes32"},{"name":"encryptedVotes","type":"bytes32"}],"name":"voteForCandidate","outputs":[],"payable":true,"stateMutability":"payable","type":"function"},{"constant":false,"inputs":[{"name":"encryptedVotes_input","type":"bytes32"}],"name":"addVotesData","outputs":[{"name":"success","type":"uint256"}],"payable":false,"stateMutability":"nonpayable","type":"function"},{"constant":true,"inputs":[{"name":"","type":"uint256"}],"name":"candidateList","outputs":[{"name":"","type":"bytes32"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[{"name":"","type":"address"}],"name":"voterInfo","outputs":[{"name":"voterAddress","type":"address"},{"name":"voted","type":"bool"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[],"name":"winningCandidate","outputs":[{"name":"winner","type":"bytes32"}],"payable":false,"stateMutability":"view","type":"function"},{"inputs":[{"name":"candidateNames","type":"bytes32[]"}],"payable":false,"stateMutability":"nonpayable","type":"constructor"}])

var pollingContractAddress= '0x4213af7f24af5a50d4b0691d5f478d2d0feda160'
var pollingInstance= PollingContract.at(pollingContractAddress)

var checkVotesIfPresent= -1;
var Chance= require('chance')
chance= new Chance(Math.random);

var aesjs = require("aes-js");	

var keyForVotersToDecrypt;

/*
 * This provides the functionality for putting a vote for a particular Candidate on the Blockchain.
*/

function voteForCandidate() {

  	candidateName = $("#candidate").val();
	//alert("candidateName: "+candidateName)

	if (candidateName.length != 0){

		var vote1or0= ChooseVote();
		//alert("VOTES: "+vote1or0)

		if(vote1or0 != 7){
			
			var encryptedVote= aesEncryption(vote1or0)
			alert("EncryptedVote: "+encryptedVote)

			/* Putting Vote on the Blockchain*/
			var candaidateVote= pollingInstance.voteForCandidate(candidateName, encryptedVote, {from: web3.eth.accounts[0]});
			alert("OK: "+candaidateVote)
			pollingInstance.addVotesData(encryptedVote,{from: web3.eth.accounts[0]})
			checkVotesIfPresent= encryptedVote
			//alert("auditVotes: "+checkVotesIfPresent)

			CountDown(5)
			
		}
		else
			alert("Yes or No Vote not clicked")
	}
	else
		alert("Candidate Name is NULL")
}

/*
 * This function provides whether the user.voter has chosen to vote a yesVote or a noVote for a particular Candidate.
*/
function ChooseVote() {

	var voteRes;
	if(document.getElementById('zeroVote').checked){
		//alert(document.getElementById('zeroVote').value)
		voteRes= document.getElementById('zeroVote').value
	}
	else if(document.getElementById('oneVote').checked){
		//alert(document.getElementById('oneVote').value)
		voteRes= document.getElementById('oneVote').value
	}
	else{
		voteRes= 7
	}
	return voteRes
}

/*
 * This is For Counting the total Votes got by each of the Candidates.
*/
function totalVotesFor(){

	candidateNames = Object.keys(candidates);
  	for (var i = 0; i < candidateNames.length; i++) {
    	let name = candidateNames[i];

		var val= countVotesFor(name)
    	$("#" + candidates[name]).html(val);
  	}
}

/*
 * This function is called for each of the Candidate to count his/her total votes, from the Blockchain.
*/

function countVotesFor(name){

	var orgString= []
	var decryptedVote= []

	var val= pollingInstance.totalVotesFor.call(name).toString()
	//alert(val)
	
	var str_array = val.split(',');

	for(var i = 0; i < str_array.length; i++) {
		// Trim the excess whitespace.
		str_array[i] = str_array[i].replace(/^\s*/, "").replace(/\s*$/, "");
		//alert("Splitted String: "+str_array[i]);

		orgString.push(web3.toAscii(str_array[i]))	
	}
	
	for(var i = 0; i < str_array.length; i++){

		//alert("OrgString: "+orgString[i])
		decryptedVote[i]= aesDecryption(orgString[i]);
	}

	var countVotes= 0;
	for(var j= 0; j < decryptedVote.length; j++){
		
		if(decryptedVote[j] == 'one')
			countVotes += 1;
	}
	//alert("VotesCounted: "+countVotes)
	return countVotes
}

function cleanArray(actual) {
  var newArray = new Array();
  for (var i = 0; i < actual.length; i++) {
    if (actual[i]) {
      newArray.push(actual[i]);
    }
  }
  return newArray;
}


function showToButtonForVotesSummary(){

	var votesCountButton= document.getElementById("votesSumm").disabled= false;
}


/*
 * This function Encrypts the Vote using AES before the votes is put on the Blockchain.
*/
function aesEncryption(text){

	// A 128-bit private key
	var key = [ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16 ];
	keyForVotersToDecrypt= key;
	// The initialization vector (must be 16 bytes)
	var iv = [ 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,35, 36 ];

	// Convert text to bytes
	var textBytes = aesjs.utils.utf8.toBytes(text);

	var aesOfb = new aesjs.ModeOfOperation.ofb(key, iv);
	var encryptedBytes = aesOfb.encrypt(textBytes);

	//alert("EncryptedBytes: "+encryptedBytes)

	// To print or store the binary data, you may convert it to hex
	var encryptedHex = aesjs.utils.hex.fromBytes(encryptedBytes);
	//alert("EncryptedHex: "+encryptedHex);

	return encryptedHex
}

/*
 * This function Decrypts the Vote using AES before the votes is used counted for each of the Candidate.
*/
function aesDecryption(encryptedHex){

	// A 128-bit private keys
	var key = [ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16 ];
	// The initialization vector (must be 16 bytes)
	var iv = [ 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,35, 36 ];

	// When ready to decrypt the hex string, convert it back to bytes
	var encryptedBytes = aesjs.utils.hex.toBytes(encryptedHex);

	var cleanedEncryptedBytes= cleanArray(encryptedBytes);
	//alert("Cleaned Array: "+cleanedEncryptedBytes)

	// The output feedback mode of operation maintains internal state,
	// so to decrypt a new instance must be instantiated.
	var aesOfb = new aesjs.ModeOfOperation.ofb(key, iv);
	var decryptedBytes = aesOfb.decrypt(cleanedEncryptedBytes);
	//alert('hi: '+decryptedBytes)

	// Convert our bytes back into text
	var decryptedText = aesjs.utils.utf8.fromBytes(decryptedBytes);

	//alert("Decrypt: "+decryptedText);
	return decryptedText;
}

function CountDown(time){

	var interval= setInterval(function() {
		document.getElementById("timerForVotesSummary").innerHTML= --time
		//document.getElementById("voteID").disabled= true

		if(time <= 0){

			document.getElementById("timerForVotesSummary").innerHTML= "Click Send Secret Key Button"
			document.getElementById("keyToCountVotes").disabled= false
			document.getElementById("secretKey").innerHTML= "[" +keyForVotersToDecrypt +"]";
			clearInterval(interval)
			
			x.style.display = "none";
		}
	
	},1000)
}

/*
 * This provides a check whether the vote casted is put on the Blockchain, which is used later for Counting.
*/
function CheckIfVoteCasted(){

	if(checkVotesIfPresent == -1){
		alert("U have not voted or Error")
	}
	else{
		var checkAuditVotes= pollingInstance.checkVotesData.call(checkVotesIfPresent)
		if(true == checkAuditVotes)
			alert("Ur Vote is Counted")
	}
}


